File: catkin_ws/src/ik_trajectory_tutorial/scripts/test_ik_trajectory_tutorial.py
#test client for ik_trajectory_tutorial

import roslib
roslib.load_manifest('ik_trajectory_tutorial')
import rospy
from ik_trajectory_tutorial.srv import ExecuteCartesianIKTrajectory
from geometry_msgs.msg import Pose
import time
import sys
import pdb
import tf
from gazebo_msgs.srv import GetModelState

#execute a Cartesian trajectory defined by a root frame, a list of 
#positions (x,y,z), and a list of orientations (quaternions: x,y,z,w)
def call_execute_cartesian_ik_trajectory_right(frame, positions, orientations):
    rospy.wait_for_service("execute_cartesian_ik_trajectory_right")

    #fill in the header (don't need seq)
    header = rospy.Header()
    header.frame_id = frame
    header.stamp = rospy.get_rostime()

    #fill in the poses
    poses = []
    for (position, orientation) in zip(positions, orientations):
        pose = Pose()
        pose.position.x = position[0]
        pose.position.y = position[1]
        pose.position.z = position[2]
        pose.orientation.x = orientation[0]
        pose.orientation.y = orientation[1]
        pose.orientation.z = orientation[2]
        pose.orientation.w = orientation[3]
        poses.append(pose)

    #call the service to execute the trajectory
    print ("calling execute_cartesian_ik_trajectory_right")
    try:
        s = rospy.ServiceProxy("execute_cartesian_ik_trajectory_right", \
                                   ExecuteCartesianIKTrajectory)
        resp = s(header, poses)
    except rospy.ServiceException as e:
        print ("error when calling execute_cartesian_ik_trajectory_right: %s"%e)
        return 0
    return resp.success

def call_execute_cartesian_ik_trajectory_left(frame, positions, orientations):
    rospy.wait_for_service("execute_cartesian_ik_trajectory_left")

    #fill in the header (don't need seq)
    header = rospy.Header()
    header.frame_id = frame
    header.stamp = rospy.get_rostime()

    #fill in the poses
    poses = []
    for (position, orientation) in zip(positions, orientations):
        pose = Pose()
        pose.position.x = position[0]
        pose.position.y = position[1]
        pose.position.z = position[2]
        pose.orientation.x = orientation[0]
        pose.orientation.y = orientation[1]
        pose.orientation.z = orientation[2]
        pose.orientation.w = orientation[3]
        poses.append(pose)

    #call the service to execute the trajectory
    print ("calling execute_cartesian_ik_trajectory_left")
    try:
        s = rospy.ServiceProxy("execute_cartesian_ik_trajectory_left", \
                                   ExecuteCartesianIKTrajectory)
        resp = s(header, poses)
    except rospy.ServiceException as e:
        print ("error when calling execute_cartesian_ik_trajectory_left: %s"%e)
        return 0
    return resp.success

#pretty-print list to string
def pplist(list):
    return ' '.join(['%2.3f'%x for x in list])

#print out the positions, velocities, and efforts of the right arm joints
if __name__ == "__main__":
    rospy.init_node("test_cartesian_ik_trajectory_executer")
    tf_listener = tf.TransformListener()
    time.sleep(.5) #give the transform listener time to get some frames

    rospy.wait_for_service('/gazebo/get_model_state')
    try:


        positions_r = [[.76, -.19, .75]]
        orientations_r = [[.92, -.09, 0.0, 1.0]]

        positions_l = [[.76, .19, .75]]
        orientations_l = [[.92, -.09, 0.0, 1.0]]

        # positions_l = [[.76, .19, .83], [0.59, 0.36, 0.93]]
        # orientations_l = [[.02, -.09, 0.0, 1.0], [0.65, -0.21, .38, .62]]
        success_r = call_execute_cartesian_ik_trajectory_right("/base_link", \
                positions_r, orientations_r)


        success_l = call_execute_cartesian_ik_trajectory_left("/base_link", \
                positions_l, orientations_l)

        #check the final pose
        (trans_r, rot_r) = tf_listener.lookupTransform('r_wrist_roll_link', 'base_link', rospy.Time(0))
        transformation_matrix = tf_listener.fromTranslationRotation(trans_r, rot_r)
        print(transformation_matrix)
        (trans_l, rot_l) = tf_listener.lookupTransform('base_link', 'l_wrist_roll_link', rospy.Time(0))
        print ("end Cartesian pose: trans", pplist(trans_r), "rot", pplist(rot_r))
        print ("end Cartesian pose: trans", pplist(trans_l), "rot", pplist(rot_l))

        if success_r and success_l:
            print ("trajectory succeeded!")
        else:
            print ("trajectory failed.")


    except rospy.ServiceException as e:
        rospy.logerr(f"Service call failed: {e}")

    # positions_r = [[.76, -.19, .83], [0.59, -0.36, 0.93]]
    # orientations_r = [[.02, -.09, 0.0, 1.0], [0.65, -0.21, .38, .62]]
    

    

File: catkin_ws/src/ik_trajectory_tutorial/scripts/test_ik_trajectory_tutorial.cpp
#include <ros/ros.h>
#include <tf/transform_listener.h>
#include <geometry_msgs/Pose.h>
#include <gazebo_msgs/GetModelState.h>
#include <ik_trajectory_tutorial/ExecuteCartesianIKTrajectory.h>

bool call_execute_cartesian_ik_trajectory_right(
    const std::string& frame,
    const std::vector<std::vector<double>>& positions,
    const std::vector<std::vector<double>>& orientations,
    ros::ServiceClient& client)
{
    ik_trajectory_tutorial::ExecuteCartesianIKTrajectory srv;
    srv.request.header.frame_id = frame;
    srv.request.header.stamp = ros::Time::now();

    // Fill in the poses
    for (size_t i = 0; i < positions.size(); ++i)
    {
        geometry_msgs::Pose pose;
        pose.position.x = positions[i][0];
        pose.position.y = positions[i][1];
        pose.position.z = positions[i][2];
        pose.orientation.x = orientations[i][0];
        pose.orientation.y = orientations[i][1];
        pose.orientation.z = orientations[i][2];
        pose.orientation.w = orientations[i][3];
        srv.request.poses.push_back(pose);
    }

    // Call the service to execute the trajectory
    ROS_INFO("Calling execute_cartesian_ik_trajectory_right");
    if (client.call(srv))
    {
        return srv.response.success;
    }
    else
    {
        ROS_ERROR("Error when calling execute_cartesian_ik_trajectory_right");
        return false;
    }
}

int main(int argc, char** argv)
{
    ros::init(argc, argv, "test_cartesian_ik_trajectory_executer");
    ros::NodeHandle nh;
    tf::TransformListener tfListener;
    ros::Duration(0.5).sleep(); // Give the transform listener time to get some frames

    ros::ServiceClient getModelStateClient = nh.serviceClient<gazebo_msgs::GetModelState>("/gazebo/get_model_state");
    ros::ServiceClient executeCartesianIKClientRight = nh.serviceClient<ik_trajectory_tutorial::ExecuteCartesianIKTrajectory>("execute_cartesian_ik_trajectory_right");
    ros::ServiceClient executeCartesianIKClientLeft = nh.serviceClient<ik_trajectory_tutorial::ExecuteCartesianIKTrajectory>("execute_cartesian_ik_trajectory_left");

    ros::service::waitForService("/gazebo/get_model_state");

    try
    {
        gazebo_msgs::GetModelState srv;
        srv.request.model_name = "wood_cube_10cm";
        srv.request.relative_entity_name = "world";
        getModelStateClient.call(srv);
        double x_r = srv.response.pose.position.x;
        double y_r = srv.response.pose.position.y;

        srv.request.model_name = "wood_cube_10cm_clone";
        getModelStateClient.call(srv);
        double x_l = srv.response.pose.position.x;
        double y_l = srv.response.pose.position.y;

        srv.request.model_name = "pr2";
        getModelStateClient.call(srv);
        double r_x = srv.response.pose.position.x;
        double r_y = srv.response.pose.position.y;

        std::vector<std::vector<double>> positions_r = {{x_r - r_x - 0.2, y_r - r_y - 0.1, 0.66}};
        std::vector<std::vector<double>> orientations_r = {{0.02, -0.09, 0.0, 1.0}};

        std::vector<std::vector<double>> positions_l = {{x_l - r_x - 0.2, y_l - r_y - 0.1, 0.66}};
        std::vector<std::vector<double>> orientations_l = {{0.02, -0.09, 0.0, 1.0}};

        bool success_r = call_execute_cartesian_ik_trajectory_right("/base_link", positions_r, orientations_r, executeCartesianIKClientRight);

        bool success_l = call_execute_cartesian_ik_trajectory_right("/base_link", positions_l, orientations_l, executeCartesianIKClientLeft);

        tf::StampedTransform transform_r;
        tfListener.lookupTransform("r_wrist_roll_link", "base_link", ros::Time(0), transform_r);
        tf::StampedTransform transform_l;
        tfListener.lookupTransform("l_wrist_roll_link", "base_link", ros::Time(0), transform_l);

        ROS_INFO("End Cartesian pose right: trans (%2.3f %2.3f %2.3f) rot (%2.3f %2.3f %2.3f %2.3f)",
                 transform_r.getOrigin().x(), transform_r.getOrigin().y(), transform_r.getOrigin().z(),
                 transform_r.getRotation().x(), transform_r.getRotation().y(), transform_r.getRotation().z(), transform_r.getRotation().w());

        ROS_INFO("End Cartesian pose left: trans (%2.3f %2.3f %2.3f) rot (%2.3f %2.3f %2.3f %2.3f)",
                 transform_l.getOrigin().x(), transform_l.getOrigin().y(), transform_l.getOrigin().z(),
                 transform_l.getRotation().x(), transform_l.getRotation().y(), transform_l.getRotation().z(), transform_l.getRotation().w());

        if (success_r && success_l)
        {
            ROS_INFO("Trajectory succeeded!");
        }
        else
        {
            ROS_INFO("Trajectory failed.");
        }
    }
    catch (ros::ServiceException& e)
    {
        ROS_ERROR("Service call failed: %s", e.what());
        return 1;
    }

    return 0;
}
